features(Tasks): Add Hardhat Tasks for Deployed Contracts (#9)
pooltogether/v4-testnet: README.md

# PoolTogeher V4 Testnet
The V4 testnet deployed contracts and essential hardhat tasks.

# Getting Started
Install `direnv` module.

We use [direnv](https://direnv.net/) to manage environment variables.  You'll likely need to install it.

```sh

> cp .envrc.example .envrv

Reviewer: ```suggestion
cp .envrc.example .envrc
```

```

To run fork scripts, deploy or perform any operation with a mainnet/testnet node you will need an Infura API key.

### Disbursement Address

> To `disburse` and `deposit` you will need to add a list of address(s) to DISBURSE_ADDRESSES.

Reviewer: ```suggestion
To `disburse` and `deposit` you will need to add a list of address(es) to DISBURSE_ADDRESSES.
```

**Example**
```.sh
export DISBURSE_ADDRESSES='0x0000000000000000000000000000000000000000,0x0000000000000000000000000000000000000001'
```

# Setup
```.sh
yarn
```

## New Deployments
```.sh
yarn deploy-rinkeby
```

```.sh
yarn deploy-mumbai
```

## Acquire Tokens & Tickets

```.sh
yarn disburse rinkeby
```

```sh
yarn disburse mumbai
```

```sh
yarn deposit rinkeby
```

```sh
yarn deposit mumbai
```

## Test Deployment Scripts

```.sh
yarn node-rinkeby
```

```.sh
yarn node-mumbai
```

# Tasks
Interact with the V4 testnet deployed contracts through hardhat tasks.

`yarn task [TASK_NAME]`

`yarn task [TASK_NAME] --wallet 0`

`yarn task [TASK_NAME] --user 0x0000000000000000000000000000000000000000`

## Calculations

> - **winningPickIndices:** Calclates a users winning picks and prize tier (distribution index) results

Reviewer: ```suggestion
- **winningPickIndices:** Calculates a user winning picks and prize tier (distribution index) results 
```

> ## DrawPrize

Reviewer: ```suggestion
## PrizeDistributor
```

> - **claim:** Claim winning draw prizes and receive tickets 

Reviewer: ```suggestion
- **claim:** Claim winning prizes and receive tickets 
```

> ## DrawHistory

Reviewer: ```suggestion
## DrawBuffer
```

> - **getDraws:** Read target draw history parameters 
> - **getOldestDraw:** Read oldest draw history parameters
> - **getNewestDraw:** Read newest draw history parameters
> - **getLiveDraws:** Read all draw history parameters between oldest and newest

Reviewer: ```suggestion
- **getDraws:** Read target draw buffer parameters 
- **getOldestDraw:** Read oldest draw buffer parameters
- **getNewestDraw:** Read newest draw buffer parameters
- **getLiveDraws:** Read all draw buffer parameters between oldest and newest
```

- **pushDraw:** Push new draw parameters
- **setDraw:** Set existing draw parameters

> ## PrizeDistributionHistory

Reviewer: ```suggestion
## PrizeDistributionBuffer
```

- **getPrizeDistribution:** Read target prize distribtion parameters
- **getOldestPrizeDistribution:** Read oldest prize distribtion parameters
- **getNewestPrizeDistribution:** Read newest prize distribtion parameters 

> - **getLivePrizeDistributionList:** Read all prize distribution history between oldest and newest

Reviewer: ```suggestion
- **getLivePrizeDistributionList:** Read all prize distribution buffer between oldest and newest
```

## PrizePool

 - **deposit:** Deposit tokens and mint tickets
 
## Ticket

 - **balanceOf:** Read balance of user address
 - **getAccountDetails:** Read account details of user address
 - **getAverageBalancesBetween:** Read average balance between epoch timestamp of user address
 - **delegate:** Delegate TWAB to delegate address
 - **transfer:** Transfer balance of ticket to recipient address

### Parameters

Certain tasks require a "user" address like `claim` and `calculate`.

> Default the `wallet` address is used. A `user` address can be passedd

Reviewer: ```suggestion
Default the `wallet` address is used. A `user` address can be passed
```
